### PR TITLE
Multihead Attention Fixes

### PR DESCRIPTION
**Pull request by OneAdder**
### Minor Changes to MHA
1. Memory allocation optimization
2. Making attention masks usable (with test)
3. Minor cleanup (unused variables etc). BTW, perhaps, add some static analyzer to CI?
4. A bit of reshuffling inside MHA forward and backward calls for better inheritance. Reason for those changes is that I'm making Llama Attention which is subclass of MHA. Source here: https://github.com/OneAdder/llm.f/blob/master/src/llama_attention.f90

### CODE REVIEW

**Comment by OneAdder** [comment]
I figured out that a little refactoring would benefit greatly. I'll add it and then this will be ready for review

File: src/nf/nf_multihead_attention.f90 [hunk 1]
```diff
@@ -68,25 +83,38 @@ end function multihead_attention_layer_cons
...
       !! General forward propagation for MultiHead Attention Mechanism
       !! Might be used for both Self and Cross Attention
       !! Self Attention: pass the same value thrice
       !! Cross Attention: pass three values for your query, key and value
       class(multihead_attention_layer), intent(in out) :: self
       real, intent(in) :: query(:, :), key(:, :), value(:, :)
+      real, optional, intent(in) :: attention_mask(:, :)
     end subroutine common_forward
 
+    pure module subroutine sdpa_forward(self, attention_mask)
```
Comment (OneAdder) [inline src/nf/nf_multihead_attention.f90:107]: Put Scaled Dot Product Attention into a separate method. This adds more flexibility.
In some cases we need to do manipulations with input projections, such as _KV Caching_ for LLama and Qwen2.

**Comment by OneAdder** [comment]
Ready for review

**Review by jvdp1 — COMMENTED** [review]
LGTM!

File: src/nf/nf_multihead_attention.f90 [hunk 2]
```diff
@@ -39,10 +39,25 @@ module nf_multihead_attention_layer
...
+    real, allocatable :: k_heads(:, :, :)
+    real, allocatable :: q_heads(:, :, :)
+    real, allocatable :: d_sdpa(:, :)
+    real, allocatable :: jacobian(:, :)
+    real, allocatable :: d_normalize(:, :, :)
   contains
 
     procedure :: common_backward
     procedure :: common_forward
+    procedure :: sdpa_forward
```
[the remark this answers is not in the document]
Comment (OneAdder) [reply src/nf/nf_multihead_attention.f90:59]: will do!

File: src/nf/nf_multihead_attention_submodule.f90 [hunk 3]
```diff
@@ -14,51 +13,91 @@ module function multihead_attention_layer_cons(n_heads) result(res)
...
-    q_heads = self % split_heads(self % query_layer % output)
+    self % d_output = self % split_heads(self % output_layer % gradient)
 
     ! iterate over heads to calculate deltas for each of them
     do concurrent(head = 1: self % n_heads)
-      dv(:, :, head) = matmul(transpose(self % attention_matrix(:, :, head)), d_output(:, :, head))
+      self % v_or_dv(:, :, head) = matmul(&
+          transpose(self % attention_matrix(:, :, head)),&
+          self % d_output(:, :, head)&
+      )
```
Comment (milancurcic) [inline src/nf/nf_multihead_attention_submodule.f90:93]: Since this PR is doing some optimizations, there's an opportunity here as well. Intrinsic `matmul` can often be slower than writing out the matrix multiplication by hand using loops, +, and -. `transpose` creates a temporary copy, and can be avoided if we're looping and assigning values explicitly. When @jvdp1 made this optimization in `dense % forward`, I think it cut the compute time in half.

File: src/nf/nf_multihead_attention_submodule.f90 [hunk 4]
```diff
@@ -14,51 +13,91 @@ module function multihead_attention_layer_cons(n_heads) result(res)
...
 
       ! calculate delta for attention matrix
-      d_sdpa = matmul(d_output(:, :, head), transpose(v_heads(:, :, head)))
+      self % d_sdpa = matmul(self % d_output(:, :, head), transpose(self % v_heads(:, :, head)))
```
Comment (milancurcic) [inline src/nf/nf_multihead_attention_submodule.f90:96]: As above, similar could be done here.

File: src/nf/nf_multihead_attention_submodule.f90 [hunk 5]
```diff
@@ -82,79 +121,19 @@ pure module subroutine common_backward(self, input, gradient)
         ! multiply output of softmax by temp jacobian matrix
         ! For computational efficiency (avoid more temp storages), scaling is also done here
         ! reshapes: [3] -> [1, 3] @ [3, 3] = [1, 3] -> [3]
-        d_normalize(seq, :, head) = reshape(matmul(&
-            reshape(d_sdpa(seq, :), [1, self % sequence_length]),&
-            jacobian * self % scaling_factor&
+        self % d_normalize(seq, :, head) = reshape(matmul(&
+            reshape(self % d_sdpa(seq, :), [1, self % sequence_length]),&
+            self % jacobian * self % scaling_factor&
```
Comment (milancurcic) [inline src/nf/nf_multihead_attention_submodule.f90:126]: As above, in this case it may be even worse since we're doing `reshape(matmul(reshape, ...))`

File: src/nf/nf_multihead_attention_submodule.f90 [hunk 6]
```diff
@@ -82,79 +121,19 @@ pure module subroutine common_backward(self, input, gradient)
...
         ), [self % sequence_length])
       end do
 
       ! calculate delta for query
-      dq(:, :, head) = matmul(d_normalize(:, :, head), k_heads(:, :, head))
+      self % q_or_dq(:, :, head) = matmul(self % d_normalize(:, :, head), self % k_heads(:, :, head))
 
       ! calculate delta for key, attention matrix should be transposed unlike for query
-      dk(:, :, head) = matmul(transpose(d_normalize(:, :, head)), q_heads(:, :, head))
+      self % k_or_dk(:, :, head) = matmul(transpose(self % d_normalize(:, :, head)), self % q_heads(:, :, head))
```
Comment (milancurcic) [inline src/nf/nf_multihead_attention_submodule.f90:134]: As above

**Review by milancurcic — APPROVED** [review]
Thanks @OneAdder this looks good, see a few comments for further optimization that we've done before in `dense % forward`. Worth trying if you don't mind, or let me know if I can help.

**Comment by milancurcic** [comment]
> Minor cleanup (unused variables etc). BTW, perhaps, add some static analyzer to CI?

I don't know how to do this and I'm too lazy to learn since I have this in my editor already :)

**Comment by OneAdder** [comment]
@milancurcic @jvdp1 Do you think replacing matmuls with custom `do concurrent` is a good idea? The issue I see here is that we are stripping the power of `-fexternal-blas` which is super fast. Perhaps, the pointers approach could be a better idea? What I was thinking: add `view` function that returns a pointer and works as `reshape` but without an extra copy. Similar to [how PyTorch does it](https://pytorch.org/docs/stable/generated/torch.Tensor.view.html)

**Comment by milancurcic** [comment]
Thank you for all the work so far, @OneAdder, and good luck with the job search! Feel free to get back to this whenever you're ready and I'll resume with the reviews of other PRs.

**Comment by milancurcic** [comment]
> @milancurcic @jvdp1 Do you think replacing matmuls with custom do concurrent is a good idea? The issue I see here is that we are stripping the power of -fexternal-blas which is super fast. Perhaps, the pointers approach could be a better idea? What I was thinking: add view function that returns a pointer and works as reshape but without an extra copy.

Indeed, it's more complicated, as you write. Since the way you have it now is correct, any performance optimization can be left for a separate PR and we can investigate in more detail what approach we should take. So I'll just go ahead an merge this PR and we can revisit the optimizations later.
